app/components/chat/ImportFolderButton.tsx

```diff
@@ -1,102 +1,73 @@
-
-          reader.onload = () => {
-            const content = reader.result as string;
-            const relativePath = file.webkitRelativePath.split('/').slice(1).join('/');
-            resolve(
-              `<boltAction type="file" filePath="${relativePath}">
-${content}
-</boltAction>`,
-            );
-          };
-          reader.onerror = reject;
-          reader.readAsText(file);
-        });
-      }),
-    );
-
-    const binaryFilesMessage =
-      binaryFiles.length > 0
-        ? `\n\nSkipped ${binaryFiles.length} binary files:\n${binaryFiles.map((f) => `- ${f}`).join('\n')}`
-        : '';
+  const [isLoading, setIsLoading] = useState(false);
 
-    const message: Message = {
-      role: 'assistant',
-      content: `I'll help you set up these files.${binaryFilesMessage}
+  const handleFileChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    const allFiles = Array.from(e.target.files || []);
 
-<boltArtifact id="imported-files" title="Imported Files" type="bundled">
-${fileArtifacts.join('\n\n')}
-</boltArtifact>`,
-      id: generateId(),
-      createdAt: new Date(),
-    };
-
-    const userMessage: Message = {
-      role: 'user',
-      id: generateId(),
-      content: 'Import my files',
-      createdAt: new Date(),
-    };
-
-    const description = `Folder Import: ${files[0].webkitRelativePath.split('/')[0]}`;
-
-    if (importChat) {
-      await importChat(description, [userMessage, message]);
+    if (allFiles.length > MAX_FILES) {
+      toast.error(
+        `This folder contains ${allFiles.length.toLocaleString()} files. This product is not yet optimized for very large projects. Please select a folder with fewer than ${MAX_FILES.toLocaleString()} files.`
+      );
+      return;
+    }
+    const folderName = allFiles[0]?.webkitRelativePath.split('/')[0] || 'Unknown Folder';
+    setIsLoading(true);
+    const loadingToast = toast.loading(`Importing ${folderName}...`);
+
+    try {
+      const filteredFiles = allFiles.filter((file) => shouldIncludeFile(file.webkitRelativePath));
+
+      if (filteredFiles.length === 0) {
+        toast.error('No files found in the selected folder');
+        return;
+      }
+
+      const fileChecks = await Promise.all(
+        filteredFiles.map(async (file) => ({
+          file,
+          isBinary: await isBinaryFile(file),
+        })),
+      );
+
+      const textFiles = fileChecks.filter((f) => !f.isBinary).map((f) => f.file);
+      const binaryFilePaths = fileChecks
+        .filter((f) => f.isBinary)
+        .map((f) => f.file.webkitRelativePath.split('/').slice(1).join('/'));
+
+      if (textFiles.length === 0) {
+        toast.error('No text files found in the selected folder');
+        return;
+      }
+
+      if (binaryFilePaths.length > 0) {
+        toast.info(`Skipping ${binaryFilePaths.length} binary files`);
+      }
+
+      const { userMessage, assistantMessage } = await createChatFromFolder(textFiles, binaryFilePaths, folderName);
```

Comment:
can this return a list of array of messages instead of object, so that we can directly pass them.
this will help if we want to return multiple messages instead of one from user and one from assistant

I like the shell command to be in a separate message with its own artifact. so that I can bundle the files artifact into one single action.. its already present the main repo